File: vscode-extension/src/ide/settings/configurables/CodexConfigurable.ts
/**
 * Codex 配置逻辑
 * 
 * 翻译自: jetbrains-plugin/src/main/kotlin/com/asakii/settings/CodexConfigurable.kt
 * 
 * 负责管理 Codex 相关的设置:
 * - 默认权限 (bypass permissions, auto cleanup contexts)
 * - 运行时设置 (Codex 路径, Web 搜索)
 * - 模型设置 (默认模型, 自定义模型)
 * - 会话默认值 (reasoning effort, summary, sandbox mode)
 */

import * as vscode from 'vscode'

// ============================================================================
// Types
// ============================================================================

/** 模型信息 */
export interface CodexModelInfo {
  displayName: string
  modelId: string
  isBuiltIn: boolean
}

/** 自定义模型配置 */
export interface CodexCustomModelConfig {
  id: string
  displayName: string
  modelId: string
}

/** 沙箱选项 */
export interface SandboxOption {
  id: string
  label: string
}

// ============================================================================
// Constants
// ============================================================================

/** 内置 Codex 模型列表 */
export const CODEX_BUILT_IN_MODELS: CodexModelInfo[] = [
  { displayName: 'GPT-5.2 Codex', modelId: 'gpt-5.2-codex', isBuiltIn: true },
  { displayName: 'GPT-5.2', modelId: 'gpt-5.2', isBuiltIn: true },
  { displayName: 'O3', modelId: 'o3', isBuiltIn: true },
  { displayName: 'O4-mini', modelId: 'o4-mini', isBuiltIn: true }
]

/** Reasoning effort 选项 */
export const REASONING_EFFORT_OPTIONS = ['minimal', 'low', 'medium', 'high', 'xhigh'] as const
export type ReasoningEffort = typeof REASONING_EFFORT_OPTIONS[number]

/** Reasoning summary 选项 */
export const REASONING_SUMMARY_OPTIONS = ['auto', 'concise', 'detailed', 'none'] as const
export type ReasoningSummary = typeof REASONING_SUMMARY_OPTIONS[number]

/** Sandbox mode 选项 */
export const SANDBOX_MODE_OPTIONS: SandboxOption[] = [
  { id: 'read-only', label: 'Chat' },
  { id: 'workspace-write', label: 'Agent' },
  { id: 'danger-full-access', label: 'Agent (full access)' }
]
export type SandboxMode = 'read-only' | 'workspace-write' | 'danger-full-access'

// ============================================================================
// Configuration Service
// ============================================================================

/**
 * Codex 配置服务
 * 
 * 提供对 VS Code 设置的统一访问接口
 */
export class CodexConfigurable {
  private static readonly SECTION = 'claudeCodePlus.codex'

  /**
   * 获取配置对象
   */
  private static getConfig(): vscode.WorkspaceConfiguration {
    return vscode.workspace.getConfiguration(this.SECTION)
  }

  // ========== Runtime Settings ==========

  /**
   * 获取 Codex 路径
   */
  static getCodexPath(): string {
    return this.getConfig().get<string>('path', '')
  }

  /**
   * 设置 Codex 路径
   */
  static async setCodexPath(path: string): Promise<void> {
    await this.getConfig().update('path', path, vscode.ConfigurationTarget.Global)
  }

  /**
   * 获取是否启用 Web 搜索
   */
  static getWebSearchEnabled(): boolean {
    return this.getConfig().get<boolean>('webSearchEnabled', false)
  }

  /**
   * 设置是否启用 Web 搜索
   */
  static async setWebSearchEnabled(enabled: boolean): Promise<void> {
    await this.getConfig().update('webSearchEnabled', enabled, vscode.ConfigurationTarget.Global)
  }

  // ========== Model Settings ==========

  /**
   * 获取默认模型 ID
   */
  static getDefaultModelId(): string {
    return this.getConfig().get<string>('defaultModelId', 'gpt-5.2-codex')
  }

  /**
   * 设置默认模型 ID
   */
  static async setDefaultModelId(modelId: string): Promise<void> {
    await this.getConfig().update('defaultModelId', modelId, vscode.ConfigurationTarget.Global)
  }

  /**
   * 获取自定义模型列表
   */
  static getCustomModels(): CodexCustomModelConfig[] {
    const models = this.getConfig().get<{ displayName: string; modelId: string }[]>('customModels', [])
    return models.map((m, i) => ({
      id: `custom_${i}_${m.modelId}`,
      displayName: m.displayName,
      modelId: m.modelId
    }))
  }

  /**
   * 设置自定义模型列表
   */
  static async setCustomModels(models: CodexCustomModelConfig[]): Promise<void> {
    const configModels = models.map(m => ({
      displayName: m.displayName,
      modelId: m.modelId
    }))
    await this.getConfig().update('customModels', configModels, vscode.ConfigurationTarget.Global)
  }

  /**
   * 获取所有可用模型 (内置 + 自定义)
   */
  static getAllCodexModels(): CodexModelInfo[] {
    const custom = this.getCustomModels().map(m => ({
      displayName: m.displayName,
      modelId: m.modelId,
      isBuiltIn: false
    }))
    return [...CODEX_BUILT_IN_MODELS, ...custom]
  }

  // ========== Session Defaults ==========

  /**
   * 获取默认 reasoning effort
   */
  static getReasoningEffort(): ReasoningEffort {
    const value = this.getConfig().get<string>('reasoningEffort', 'medium')
    return this.isValidReasoningEffort(value) ? value : 'medium'
  }

  /**
   * 设置默认 reasoning effort
   */
  static async setReasoningEffort(effort: ReasoningEffort): Promise<void> {
    await this.getConfig().update('reasoningEffort', effort, vscode.ConfigurationTarget.Global)
  }

  /**
   * 获取默认 reasoning summary
   */
  static getReasoningSummary(): ReasoningSummary {
    const value = this.getConfig().get<string>('reasoningSummary', 'auto')
    return this.isValidReasoningSummary(value) ? value : 'auto'
  }

  /**
   * 设置默认 reasoning summary
   */
  static async setReasoningSummary(summary: ReasoningSummary): Promise<void> {
    await this.getConfig().update('reasoningSummary', summary, vscode.ConfigurationTarget.Global)
  }

  /**
   * 获取默认 sandbox mode
   */
  static getSandboxMode(): SandboxMode {
    const value = this.getConfig().get<string>('sandboxMode', 'workspace-write')
    return this.isValidSandboxMode(value) ? value : 'workspace-write'
  }

  /**
   * 设置默认 sandbox mode
   */
  static async setSandboxMode(mode: SandboxMode): Promise<void> {
    await this.getConfig().update('sandboxMode', mode, vscode.ConfigurationTarget.Global)
  }

  /**
   * 获取 sandbox mode 标签
   */
  static getSandboxModeLabel(mode: SandboxMode): string {
    const option = SANDBOX_MODE_OPTIONS.find(o => o.id === mode)
    return option?.label ?? mode
  }

  // ========== Auto Cleanup ==========

  /**
   * 获取默认自动清理上下文设置
   */
  static getDefaultAutoCleanupContexts(): boolean {
    return this.getConfig().get<boolean>('defaultAutoCleanupContexts', false)
  }

  /**
   * 设置默认自动清理上下文设置
   */
  static async setDefaultAutoCleanupContexts(value: boolean): Promise<void> {
    await this.getConfig().update('defaultAutoCleanupContexts', value, vscode.ConfigurationTarget.Global)
  }

  // ========== Validation ==========

  /**
   * 验证模型 ID 是否有效
   */
  static isValidModelId(modelId: string): boolean {
    if (!modelId || modelId.trim().length === 0) {
      return false
    }
    return CODEX_BUILT_IN_MODELS.some(m => m.modelId === modelId) ||
           this.getCustomModels().some(m => m.modelId === modelId)
  }

  /**
   * 验证 reasoning effort 是否有效
   */
  static isValidReasoningEffort(effort: string): effort is ReasoningEffort {
    return REASONING_EFFORT_OPTIONS.includes(effort as ReasoningEffort)
  }

  /**
   * 验证 reasoning summary 是否有效
   */
  static isValidReasoningSummary(summary: string): summary is ReasoningSummary {
    return REASONING_SUMMARY_OPTIONS.includes(summary as ReasoningSummary)
  }

  /**
   * 验证 sandbox mode 是否有效
   */
  static isValidSandboxMode(mode: string): mode is SandboxMode {
    return SANDBOX_MODE_OPTIONS.some(o => o.id === mode)
  }

  // ========== Change Listeners ==========

  /**
   * 监听配置变更
   */
  static onDidChangeConfiguration(
    callback: (e: vscode.ConfigurationChangeEvent) => void
  ): vscode.Disposable {
    return vscode.workspace.onDidChangeConfiguration(e => {
      if (e.affectsConfiguration('claudeCodePlus.codex')) {
        callback(e)
      }
    })
  }
}

// ============================================================================
// Exports
// ============================================================================

export default CodexConfigurable

File: jetbrains-plugin/src/main/kotlin/com/asakii/settings/CodexConfigurable.kt
package com.asakii.settings

import com.asakii.plugin.compat.BrowseButtonCompat
import com.intellij.openapi.fileChooser.FileChooserDescriptorFactory
import com.intellij.openapi.options.SearchableConfigurable
import com.intellij.openapi.ui.ComboBox
import com.intellij.openapi.ui.TextFieldWithBrowseButton
import com.intellij.ui.components.JBCheckBox
import com.intellij.ui.components.JBLabel
import com.intellij.ui.components.JBScrollPane
import com.intellij.ui.components.JBTextField
import com.intellij.ui.dsl.builder.Align
import com.intellij.ui.dsl.builder.AlignX
import com.intellij.ui.dsl.builder.panel
import com.intellij.ui.table.JBTable
import javax.swing.DefaultComboBoxModel
import javax.swing.JButton
import javax.swing.JComponent
import javax.swing.JOptionPane
import javax.swing.JPanel
import javax.swing.ListSelectionModel
import javax.swing.table.DefaultTableModel
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import java.awt.Insets

class CodexConfigurable : SearchableConfigurable {

    private var mainPanel: JPanel? = null
    private var codexPathField: TextFieldWithBrowseButton? = null
    private var webSearchCheckBox: JBCheckBox? = null
    private var defaultBypassPermissionsCheckbox: JBCheckBox? = null
    private var defaultAutoCleanupContextsCheckbox: JBCheckBox? = null
    private var defaultModelCombo: ComboBox<ModelInfo>? = null
    private var defaultReasoningEffortCombo: ComboBox<String>? = null
    private var defaultReasoningSummaryCombo: ComboBox<String>? = null
    private var defaultSandboxModeCombo: ComboBox<SandboxOption>? = null
    private var customModelsTable: JBTable? = null
    private var customModelsTableModel: DefaultTableModel? = null
    private var addModelButton: JButton? = null
    private var editModelButton: JButton? = null
    private var removeModelButton: JButton? = null

    override fun getId(): String = "com.asakii.settings.codex"

    override fun getDisplayName(): String = "Codex"

    override fun createComponent(): JComponent {
        initComponents()
        
        mainPanel = panel {
            group("Default Permissions") {
                row {
                    cell(defaultBypassPermissionsCheckbox!!)
                }
                row {
                    comment("Skip confirmation dialogs for file edits and bash commands. Use with caution.")
                }
                row {
                    cell(defaultAutoCleanupContextsCheckbox!!)
                }
                row {
                    comment("Enabled contexts are cleared after send; disabled contexts stay.")
                }
            }

            group("Runtime Settings") {
                row("Codex path:") {
                    cell(codexPathField!!)
                        .align(AlignX.FILL)
                        .resizableColumn()
                        .comment("Path to Codex executable. Leave empty to auto-detect from system PATH.")
                }
                row {
                    cell(webSearchCheckBox!!)
                        .comment("Allow Codex to request web searches (features.web_search_request).")
                }
            }

            group("Model Settings") {
                row("Default model:") {
                    cell(defaultModelCombo!!)
                        .comment("gpt-5.4 = Latest | gpt-5.3-codex = Codex optimized")
                }
            }

            collapsibleGroup("Custom Models") {
                row {
                    cell(JBScrollPane(customModelsTable))
                        .align(Align.FILL)
                }.resizableRow()
                row {
                    cell(addModelButton!!)
                    cell(editModelButton!!)
                    cell(removeModelButton!!)
                }
            }

            group("Session Defaults") {
                row("Reasoning effort:") {
                    cell(defaultReasoningEffortCombo!!)
                        .comment("Controls reasoning depth for Codex responses.")
                }
                row("Reasoning summary:") {
                    cell(defaultReasoningSummaryCombo!!)
                        .comment("Summary style for reasoning output when supported.")
                }
                row("Sandbox mode:") {
                    cell(defaultSandboxModeCombo!!)
                        .comment("Controls file system and network access permissions.")
                }
            }
        }

        reset()
        return mainPanel!!
    }

    private fun initComponents() {
        // Default Permissions
        defaultBypassPermissionsCheckbox = JBCheckBox("Default bypass permissions").apply {
            toolTipText = "When enabled, new sessions will automatically use bypass permissions mode"
        }
        defaultAutoCleanupContextsCheckbox = JBCheckBox("Default auto cleanup contexts").apply {
            toolTipText = "When enabled, new sessions will auto clear enabled contexts after sending"
        }

        // Runtime Settings
        val descriptor = FileChooserDescriptorFactory.createSingleFileNoJarsDescriptor()
        codexPathField = TextFieldWithBrowseButton().apply {
            BrowseButtonCompat.addBrowseFolderListener(
                this,
                "Select Codex Executable",
                "Choose the path to codex executable",
                null,
                descriptor
            )
            toolTipText = "Leave empty to auto-detect from system PATH"
            (textField as? JBTextField)?.let { tf ->
                tf.emptyText.text = "Detecting Codex..."
                com.intellij.openapi.application.ApplicationManager.getApplication().executeOnPooledThread {
                    val codexInfo = AgentSettingsService.detectCodexInfo()
                    com.intellij.openapi.application.ApplicationManager.getApplication().invokeLater {
                        tf.emptyText.text = codexInfo?.let {
                            if (it.version != null) "${it.path} (${it.version})" else it.path
                        } ?: "Auto-detect from system PATH (Codex not found)"
                    }
                }
            }
        }

        webSearchCheckBox = JBCheckBox("Enable web search")

        // Model Settings
        defaultModelCombo = ComboBox<ModelInfo>().apply {
            renderer = ModelInfoRenderer()
            toolTipText = "Default model for new Codex sessions"
        }

        // Custom Models Table
        customModelsTableModel = object : DefaultTableModel(arrayOf("Display Name", "Model ID"), 0) {
            override fun isCellEditable(row: Int, column: Int) = false
        }
        customModelsTable = JBTable(customModelsTableModel).apply {
            setSelectionMode(ListSelectionModel.SINGLE_SELECTION)
            tableHeader.reorderingAllowed = false
        }

        addModelButton = JButton("Add").apply {
            toolTipText = "Add a new custom model"
        }
        editModelButton = JButton("Edit").apply {
            toolTipText = "Edit selected custom model"
            isEnabled = false
        }
        removeModelButton = JButton("Remove").apply {
            toolTipText = "Remove selected custom model"
            isEnabled = false
        }

        customModelsTable!!.selectionModel.addListSelectionListener {
            val hasSelection = customModelsTable!!.selectedRow >= 0
            editModelButton?.isEnabled = hasSelection
            removeModelButton?.isEnabled = hasSelection
        }

        addModelButton!!.addActionListener {
            showCustomModelDialog(null, null) { displayName, modelId ->
                customModelsTableModel?.addRow(arrayOf(displayName, modelId))
                refreshModelCombo()
            }
        }
        editModelButton!!.addActionListener {
            val selectedRow = customModelsTable!!.selectedRow
            if (selectedRow >= 0) {
                val currentName = customModelsTableModel?.getValueAt(selectedRow, 0) as? String ?: ""
                val currentModelId = customModelsTableModel?.getValueAt(selectedRow, 1) as? String ?: ""
                showCustomModelDialog(currentName, currentModelId) { displayName, modelId ->
                    customModelsTableModel?.setValueAt(displayName, selectedRow, 0)
                    customModelsTableModel?.setValueAt(modelId, selectedRow, 1)
                    refreshModelCombo()
                }
            }
        }
        removeModelButton!!.addActionListener {
            val selectedRow = customModelsTable!!.selectedRow
            if (selectedRow >= 0) {
                customModelsTableModel?.removeRow(selectedRow)
                refreshModelCombo()
            }
        }

        // Session Defaults
        defaultReasoningEffortCombo = ComboBox(
            DefaultComboBoxModel(arrayOf("minimal", "low", "medium", "high", "xhigh"))
        ).apply {
            toolTipText = "Default reasoning effort for new Codex sessions"
        }

        defaultReasoningSummaryCombo = ComboBox(
            DefaultComboBoxModel(arrayOf("auto", "concise", "detailed", "none"))
        ).apply {
            toolTipText = "Default reasoning summary behavior for Codex sessions"
        }

        defaultSandboxModeCombo = ComboBox(
            DefaultComboBoxModel(
                arrayOf(
                    SandboxOption("read-only", "Chat"),
                    SandboxOption("workspace-write", "Agent"),
                    SandboxOption("danger-full-access", "Agent (full access)")
                )
            )
        ).apply {
            renderer = SandboxOptionRenderer()
            toolTipText = "Default sandbox mode for new Codex sessions"
        }

        refreshModelCombo()
    }

    override fun isModified(): Boolean {
        val settings = AgentSettingsService.getInstance()
        return codexPathField?.text?.trim() != settings.codexPath ||
            (webSearchCheckBox?.isSelected ?: false) != settings.codexWebSearchEnabled ||
            (defaultBypassPermissionsCheckbox?.isSelected ?: false) != settings.defaultBypassPermissions ||
            (defaultAutoCleanupContextsCheckbox?.isSelected ?: false) != settings.codexDefaultAutoCleanupContexts ||
            (defaultModelCombo?.selectedItem as? ModelInfo)?.modelId != settings.codexDefaultModelId ||
            defaultReasoningEffortCombo?.selectedItem != settings.codexDefaultReasoningEffort ||
            defaultReasoningSummaryCombo?.selectedItem != settings.codexDefaultReasoningSummary ||
            (defaultSandboxModeCombo?.selectedItem as? SandboxOption)?.id != settings.codexDefaultSandboxMode ||
            getCustomModelsFromTable().map { it.displayName to it.modelId } != settings.getCodexCustomModels().map { it.displayName to it.modelId }
    }

    override fun apply() {
        val settings = AgentSettingsService.getInstance()
        settings.codexPath = codexPathField?.text?.trim() ?: ""
        settings.codexWebSearchEnabled = webSearchCheckBox?.isSelected ?: false
        settings.defaultBypassPermissions = defaultBypassPermissionsCheckbox?.isSelected ?: false
        settings.codexDefaultAutoCleanupContexts = defaultAutoCleanupContextsCheckbox?.isSelected ?: false
        val selectedModel = defaultModelCombo?.selectedItem as? ModelInfo
        settings.codexDefaultModelId = selectedModel?.modelId ?: settings.getCodexBuiltInModels().first().modelId
        settings.codexDefaultReasoningEffort = defaultReasoningEffortCombo?.selectedItem as? String ?: "medium"
        settings.codexDefaultReasoningSummary = defaultReasoningSummaryCombo?.selectedItem as? String ?: "auto"
        settings.codexDefaultSandboxMode =
            (defaultSandboxModeCombo?.selectedItem as? SandboxOption)?.id ?: "workspace-write"
        val existingModels = settings.getCodexCustomModels().associateBy { it.modelId }
        val customModels = getCustomModelsFromTable().map { model ->
            val existing = existingModels[model.modelId]
            val id = existing?.id ?: "custom_${System.currentTimeMillis()}_${model.modelId.hashCode().toUInt()}"
            CustomModelConfig(id, model.displayName, model.modelId)
        }
        settings.setCodexCustomModels(customModels)
        settings.notifyChange()
    }

    override fun reset() {
        val settings = AgentSettingsService.getInstance()
        codexPathField?.text = settings.codexPath
        webSearchCheckBox?.isSelected = settings.codexWebSearchEnabled
        defaultBypassPermissionsCheckbox?.isSelected = settings.defaultBypassPermissions
        defaultAutoCleanupContextsCheckbox?.isSelected = settings.codexDefaultAutoCleanupContexts
        customModelsTableModel?.rowCount = 0
        settings.getCodexCustomModels().forEach { model ->
            customModelsTableModel?.addRow(arrayOf(model.displayName, model.modelId))
        }
        refreshModelCombo()
        val savedDefaultModelId = settings.codexDefaultModelId
        val allModels = settings.getAllCodexModels()
        val matchingModel = allModels.find { it.modelId == savedDefaultModelId }
        if (matchingModel != null) {
            defaultModelCombo?.selectedItem = matchingModel
        } else {
            defaultModelCombo?.selectedItem = allModels.firstOrNull { it.isBuiltIn }
        }
        defaultReasoningEffortCombo?.selectedItem = settings.codexDefaultReasoningEffort
        defaultReasoningSummaryCombo?.selectedItem = settings.codexDefaultReasoningSummary
        val sandboxMode = settings.codexDefaultSandboxMode
        val sandboxModel = defaultSandboxModeCombo?.model as? DefaultComboBoxModel<*>
        val sandboxOptions = sandboxModel?.let { model ->
            (0 until model.size).mapNotNull { idx ->
                model.getElementAt(idx) as? SandboxOption
            }
        } ?: emptyList()
        val sandboxSelection = sandboxOptions.find { it.id == sandboxMode } ?: sandboxOptions.firstOrNull()
        if (sandboxSelection != null) {
            defaultSandboxModeCombo?.selectedItem = sandboxSelection
        }
    }

    override fun disposeUIResources() {
        codexPathField = null
        webSearchCheckBox = null
        defaultBypassPermissionsCheckbox = null
        defaultAutoCleanupContextsCheckbox = null
        defaultModelCombo = null
        defaultReasoningEffortCombo = null
        defaultReasoningSummaryCombo = null
        defaultSandboxModeCombo = null
        customModelsTable = null
        customModelsTableModel = null
        addModelButton = null
        editModelButton = null
        removeModelButton = null
        mainPanel = null
    }

    private data class SandboxOption(val id: String, val label: String) {
        override fun toString(): String = label
    }

    private class SandboxOptionRenderer : javax.swing.DefaultListCellRenderer() {
        override fun getListCellRendererComponent(
            list: javax.swing.JList<*>?,
            value: Any?,
            index: Int,
            isSelected: Boolean,
            cellHasFocus: Boolean
        ): java.awt.Component {
            val component = super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus)
            if (value is SandboxOption) {
                text = value.label
            }
            return component
        }
    }

    private fun refreshModelCombo() {
        val currentSelection = defaultModelCombo?.selectedItem as? ModelInfo
        val settings = AgentSettingsService.getInstance()
        val allModels = settings.getCodexBuiltInModels() + getCustomModelsFromTable()
        defaultModelCombo?.model = DefaultComboBoxModel(allModels.toTypedArray())

        if (currentSelection != null) {
            val matchingModel = allModels.find { it.modelId == currentSelection.modelId }
            if (matchingModel != null) {
                defaultModelCombo?.selectedItem = matchingModel
            }
        }
    }

    private fun getCustomModelsFromTable(): List<ModelInfo> {
        val tableModel = customModelsTableModel ?: return emptyList()
        val result = mutableListOf<ModelInfo>()
        for (i in 0 until tableModel.rowCount) {
            val displayName = tableModel.getValueAt(i, 0) as? String ?: continue
            val modelId = tableModel.getValueAt(i, 1) as? String ?: continue
            if (displayName.isBlank() || modelId.isBlank()) continue
            result.add(ModelInfo(modelId = modelId, displayName = displayName, isBuiltIn = false))
        }
        return result
    }

    private fun showCustomModelDialog(
        currentDisplayName: String?,
        currentModelId: String?,
        onSave: (displayName: String, modelId: String) -> Unit
    ) {
        val isEdit = currentDisplayName != null

        val dialogPanel = JPanel(GridBagLayout())
        val gbc = GridBagConstraints().apply {
            insets = Insets(4, 4, 4, 4)
            anchor = GridBagConstraints.WEST
        }

        gbc.gridx = 0
        gbc.gridy = 0
        dialogPanel.add(JBLabel("Display Name:"), gbc)

        val displayNameField = JBTextField(20).apply {
            text = currentDisplayName ?: ""
            toolTipText = "Name shown in the model selector (e.g., 'My Codex Model')"
        }
        gbc.gridx = 1
        gbc.fill = GridBagConstraints.HORIZONTAL
        gbc.weightx = 1.0
        dialogPanel.add(displayNameField, gbc)

        gbc.gridx = 0
        gbc.gridy = 1
        gbc.fill = GridBagConstraints.NONE
        gbc.weightx = 0.0
        dialogPanel.add(JBLabel("Model ID:"), gbc)

        val modelIdField = JBTextField(30).apply {
            text = currentModelId ?: ""
            toolTipText = "Codex model ID (e.g., 'gpt-5.4')"
        }
        gbc.gridx = 1
        gbc.fill = GridBagConstraints.HORIZONTAL
        gbc.weightx = 1.0
        dialogPanel.add(modelIdField, gbc)

        gbc.gridx = 0
        gbc.gridy = 2
        gbc.gridwidth = 2
        gbc.fill = GridBagConstraints.HORIZONTAL
        dialogPanel.add(JBLabel("<html><font color='gray' size='-1'>Model ID examples: gpt-5.4, gpt-5.2</font></html>"), gbc)

        val result = JOptionPane.showConfirmDialog(
            mainPanel,
            dialogPanel,
            if (isEdit) "Edit Custom Model" else "Add Custom Model",
            JOptionPane.OK_CANCEL_OPTION,
            JOptionPane.PLAIN_MESSAGE
        )

        if (result == JOptionPane.OK_OPTION) {
            val displayName = displayNameField.text.trim()
            val modelId = modelIdField.text.trim()

            if (displayName.isNotEmpty() && modelId.isNotEmpty()) {
                onSave(displayName, modelId)
            } else {
                JOptionPane.showMessageDialog(
                    mainPanel,
                    "Both Display Name and Model ID are required.",
                    "Validation Error",
                    JOptionPane.WARNING_MESSAGE
                )
            }
        }
    }
}
